### PR TITLE
[BUGFIX] Check outputs too before returning.

### PR DESCRIPTION
**Pull request by srkreddy1238**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

File: nnvm/src/compiler/precompute_prune.cc [hunk 1]
```diff
@@ -75,17 +75,17 @@ nnvm::Graph PrecomputePrune(nnvm::Graph src) {
...
-  }
-
   for (auto& e : src.outputs) {
     if (pruned.count(e.node.get())) {
       e = replace_pruned_entry(e);
     }
   }
 
+  // nothing being pruned.
+  if (non_var_edge == 0) {
```
Comment (reminisce) [inline nnvm/src/compiler/precompute_prune.cc]: In some cases, this return is still too early. For example, `replace_pruned_entry` may still be called to replace a parameter node and modify its name by appending a version number. This change of the parameter name must be propagated to the frontend by setting the attribute "output_names" to the graph. Therefore, simply removing this block would solve the issue.

**Comment by srkreddy1238** [comment]
@reminisce 
What you said is right. But, I have seen some test cases failure with complete removal. Reason being a test case with no inputs(all params) results in no functions for llvm. I made changes for the failure test cases to let at least one input.

Conclusion here would be "we may not be able to have a graph compiled with precomute_prune pass  while no inputs (all params)". Though its not a realistic having a graph with no inputs.

cc @tqchen please advice on this change.

**Comment by reminisce** [comment]
@srkreddy1238 You are right, I encountered the same thing and added some special handling in compiler and runtime in my private branch. I should be able to upstream the work in a month.

**Comment by tqchen** [comment]
The main principle of passes as they should be conservative and be able to handle all the cases, so it is fine to not being able to prune graph with no variables. This problem can be resolved on NNVMv2(relay) by using a more formal constant folder.

Thanks @reminisce for reviewing, can you explicitly approve or add more reviews?

**Comment by reminisce** [comment]
Thank @tqchen for the explanation.

@srkreddy1238 While removing the early return solves the issue to a certain level, I'm worried that tweaking the `opt_level`to 1 in some unit tests may skip testing some important passes, such as operator fusion.

A quick fix for the problem of no functions for LLVM can be passing a graph attribute to the frontend indicating whether the `pre_graph` needs to be compiled and run, or just simply update the param names in the function `precompute_prune`.

The more correct way of fixing this should be adding special case handling in the compiler and runtime when no functions can be fed to LLVM.

**Comment by srkreddy1238** [comment]
Yeah, I though of other passed being ineffective with opt_level=1.
pre_graph option is too much effort to test scripts while finding exact test cases need this flag.

I think handling in LLVM to make a dummy module is an ideal option.

**Comment by reminisce** [comment]
For the first method, you don't need to change any test scripts. The root cause of failure comes from compiling a `pre_graph` with pure variable nodes. You just need to use the attribute to decide whether to skip compiling the `pre_graph` or not.

Anyway, the second method should be the most correct way of fixing this issue. I have it in my branch and can upstream it in a month if that's not too late.

**Comment by srkreddy1238** [comment]
@reminisce welcome to share the patch by handling in LLVM which is a clean fix for this issue.
Hence closing this PR here, further discussion can be on issue tracker #1676

**Pull request closed without merging**